### PR TITLE
feat(tradeshield): add tradeshield commands

### PR DESCRIPTION
**Pull request by cosmic-vagabond**
### What has Changed?

* add spot order create order CLI
* add perpetual open limit/market order CLI
* add perpetual close order CLI
* add new test coverage
* market orders are executed directly
* added new CLI for cancel single order for both spot and perpetual

#### Create spot order

```
$ elysd tx tradeshield create-spot-order limitbuy 1000000ibc/2180E84E20F5679FCC760D8C165B60F42065DEF7F46A72B447CFF1B7DC6C0A65 ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4 0.5 --from=validator --fees 1000000uelys --gas=auto --gas-adjustment 1.6 -y
```

#### Query spot order

```
$ elysd q tradeshield list-pending-spot-order
pagination:
  next_key: null
  total: "0"
pending_spot_order:
- date: null
  order_amount:
    amount: "1000000"
    denom: ibc/2180E84E20F5679FCC760D8C165B60F42065DEF7F46A72B447CFF1B7DC6C0A65
  order_id: "1"
  order_price:
    base_denom: ibc/2180E84E20F5679FCC760D8C165B60F42065DEF7F46A72B447CFF1B7DC6C0A65
    quote_denom: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4
    rate: "0.500000000000000000"
  order_target_denom: ""
  order_type: LIMITBUY
  owner_address: elys1g3qnq7apxv964cqj0hza0pnwsw3q920lcc5lyg
  status: PENDING
```

#### Create perpetual open order

```
$ elysd tx tradeshield create-perpetual-open-order limitopen long 10 2 ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4 1000000ibc/2180E84E20F5679FCC760D8C165B60F42065DEF7F46A72B447CFF1B7DC6C0A65 0.5 --from=validator -y
```

#### Query perpetual order

```
$ elysd q tradeshield list-pending-perpetual-order                           
Error: rpc error: code = Unknown desc = rpc error: code = Internal desc = rpc error: code = NotFound desc = pool not found: unknown request
```

Error due to a regression with the accounted pool module hooks that does not create the perpetual pool anymore, issue addressed in a separate PR.

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/elys-network/elys/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/elys-network/elys/blob/main/docs/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/elys-network/elys/blob/main/CONTRIBUTING.md#testing)
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### Deployment Notes

Are there any specific considerations to take into account when deploying these changes? This may include new dependencies, scripts that need to be executed, or any aspects that can only be evaluated in a deployed environment.

### Screenshots and Videos

Please provide any relevant before and after screenshots by uploading them here. Additionally, demo videos can be highly beneficial in demonstrating the process.

### CODE REVIEW

**Review by avkr003 — APPROVED** [review]

**Review by amityadav0 — APPROVED** [review]

**Comment by amityadav0** [comment]
- Create perpetual order
```
ay@Amits-MacBook-Pro code % elysd tx tradeshield create-perpetual-open-order limitopen long 10 2 ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4 1000000ibc/2180E84E20F5679FCC760D8C165B60F42065DEF7F46A72B447CFF1B7DC6C0A65 0.5 --from=validator --gas=400000 -y
```

-Query result
```
ay@Amits-MacBook-Pro code % elysd q tradeshield list-pending-perpetual-order
pagination:
  next_key: null
  total: "0"
pending_perpetual_order:
- borrow_interest_rate: "0.000000000000000000"
  funding_rate: "0.000000000000000000"
  liquidation_price: "4.426370488615094064"
  perpetual_order:
    collateral:
      amount: "1000000"
      denom: ibc/2180E84E20F5679FCC760D8C165B60F42065DEF7F46A72B447CFF1B7DC6C0A65
    leverage: "10.000000000000000000"
    order_id: "1"
    owner_address: elys1g3qnq7apxv964cqj0hza0pnwsw3q920lcc5lyg
    perpetual_order_type: LIMITOPEN
    poolId: "2"
    position: LONG
    position_id: "0"
    status: PENDING
    stop_loss_price: "0.000000000000000000"
    take_profit_price: "10000000000000000000000000000000000000000.000000000000000000"
    trading_asset: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4
    trigger_price:
      rate: "0.500000000000000000"
      trading_asset_denom: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4
  position_size:
    amount: "2134932"
    denom: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4

```

Fixed the issue

**Comment by amityadav0** [comment]
Tested cancel order too
```
ay@Amits-MacBook-Pro code % elysd tx tradeshield cancel-perpetual-order 1 --from=validator --gas=400000 -y
code: 0
codespace: ""
data: ""
events: []
gas_used: "0"
gas_wanted: "0"
height: "0"
info: ""
logs: []
raw_log: '[]'
timestamp: ""
tx: null
txhash: 139A42587FE39CCC0C2BEC29F473B0C80C645DE06CD24915E086A5E4CA92CB31
ay@Amits-MacBook-Pro code % elysd q tradeshield list-pending-perpetual-order                             
pagination:
  next_key: null
  total: "0"
pending_perpetual_order: []
```

**Review by avkr003 — CHANGES_REQUESTED** [review]

File: x/tradeshield/types/messages_pending_perpetual_order.go [hunk 1]
```diff
@@ -6,84 +6,237 @@ import (
...
+}
+
+func (msg *MsgCreatePerpetualOpenOrder) ValidateBasic() error {
+	_, err := sdk.AccAddressFromBech32(msg.OwnerAddress)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid owner address (%s)", err)
+	}
+
+	// Validate trigger price
+	if msg.TriggerPrice == nil {
```
Comment (avkr003) [inline x/tradeshield/types/messages_pending_perpetual_order.go:71]: Trigger price can't be negative right?

File: x/tradeshield/types/messages_pending_perpetual_order.go [hunk 2]
```diff
@@ -6,84 +6,237 @@ import (
...
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "trigger price cannot be nil")
+	}
+
+	// Validate collateral
+	if !msg.Collateral.IsValid() {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "invalid collateral")
+	}
+
+	// Validate trading asset
+	if msg.TradingAsset == "" {
```
Comment (avkr003) [inline x/tradeshield/types/messages_pending_perpetual_order.go:81]: use sdk.ValidateDenom to validate denoms. there are length constraints and what characters allowed

File: x/tradeshield/types/messages_pending_perpetual_order.go [hunk 3]
```diff
@@ -6,84 +6,237 @@ import (
...
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "leverage cannot be nil or negative")
+	}
+
+	// Validate take profit price
+	if msg.TakeProfitPrice.IsNegative() {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "take profit price cannot be negative")
+	}
+
+	// Validate stop loss price
+	if msg.StopLossPrice.IsNegative() {
```
Comment (avkr003) [inline x/tradeshield/types/messages_pending_perpetual_order.go:96]: msg.StopLossPrice != nil

File: x/tradeshield/types/messages_pending_perpetual_order.go [hunk 4]
```diff
@@ -6,84 +6,237 @@ import (
...
 }
 
-func (msg *MsgCreatePendingPerpetualOrder) GetSignBytes() []byte {
+func (msg *MsgCreatePerpetualCloseOrder) GetSignBytes() []byte {
 	bz := ModuleCdc.MustMarshalJSON(msg)
 	return sdk.MustSortJSON(bz)
 }
 
-func (msg *MsgCreatePendingPerpetualOrder) ValidateBasic() error {
+func (msg *MsgCreatePerpetualCloseOrder) ValidateBasic() error {
```
Comment (avkr003) [inline x/tradeshield/types/messages_pending_perpetual_order.go:144]: Validation for trigger price

File: x/tradeshield/types/messages_pending_perpetual_order.go [hunk 5]
```diff
@@ -6,84 +6,237 @@ import (
...
 
-func (msg *MsgUpdatePendingPerpetualOrder) ValidateBasic() error {
+func (msg *MsgUpdatePerpetualOrder) ValidateBasic() error {
 	_, err := sdk.AccAddressFromBech32(msg.OwnerAddress)
 	if err != nil {
 		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid creator address (%s)", err)
 	}
+
+	// Validate trigger price
+	if msg.TriggerPrice == nil {
```
Comment (avkr003) [inline x/tradeshield/types/messages_pending_perpetual_order.go:190]: negative trigger price allowed?

File: x/tradeshield/types/messages_pending_spot_order.go [hunk 6]
```diff
@@ -6,98 +6,160 @@ import (
...
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "order price cannot be nil")
+	}
+
+	// Validate order amount
+	if !msg.OrderAmount.IsValid() {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "invalid order amount")
+	}
+
+	// Validate order target denom
+	if msg.OrderTargetDenom == "" {
```
Comment (avkr003) [inline x/tradeshield/types/messages_pending_spot_order.go:67]: same as above ValidateDenom

File: x/tradeshield/types/messages_pending_spot_order.go [hunk 7]
```diff
@@ -6,98 +6,160 @@ import (
...
 
-func (msg *MsgCreatePendingSpotOrder) ValidateBasic() error {
+func (msg *MsgCreateSpotOrder) ValidateBasic() error {
 	_, err := sdk.AccAddressFromBech32(msg.OwnerAddress)
 	if err != nil {
 		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid creator address (%s)", err)
 	}
+
+	// Validate order price
+	if msg.OrderPrice == nil {
```
Comment (avkr003) [inline x/tradeshield/types/messages_pending_spot_order.go:57]: Order price can't be negative right?

File: x/tradeshield/types/messages_pending_spot_order.go [hunk 8]
```diff
@@ -6,98 +6,160 @@ import (
...
 
-func (msg *MsgUpdatePendingSpotOrder) ValidateBasic() error {
+func (msg *MsgUpdateSpotOrder) ValidateBasic() error {
 	_, err := sdk.AccAddressFromBech32(msg.OwnerAddress)
 	if err != nil {
 		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid owner address (%s)", err)
 	}
+
+	// Validate order price
+	if msg.OrderPrice == nil {
```
Comment (avkr003) [inline x/tradeshield/types/messages_pending_spot_order.go:111]: same as above

**Review by avkr003 — CHANGES_REQUESTED** [review]

File: x/tradeshield/types/messages_pending_perpetual_order.go (hunk 3 above)
Comment (amityadav0) [reply x/tradeshield/types/messages_pending_perpetual_order.go:96]: this case is not possible, you can try with your IDE

**Review by avkr003 — APPROVED** [review]

**Pull request merged**